### PR TITLE
feat: [DEL-4240]: Add delegate group whitelisting for ng task

### PR DESCRIPTION
**Pull request by jennyjames21**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/33951)

### CODE REVIEW

**Comment by jennyjames21** [comment]
trigger ti2

**Comment by jennyjames21** [comment]
trigger ti2

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 4 of 7 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33951)* status: :shipit: complete! ✓

**Review by raghuAtWings — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33951)* status: :shipit: complete! ✓

___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java` line 89 at r2](https://reviewable.io/reviews/harness/harness-core/33951#-N3v_VjF1PJDlSpAzQqx:-N3v_VjF1PJDlSpAzQqy:boe1q6w) ([raw file](https://github.com/harness/harness-core/blob/f0ccd540460b694f166e417a51c7816d0436927e/420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java#L89)):*
> ```Java
>   private LoadingCache<ImmutablePair<String, String>, List<Delegate>> delegatesFromGroupCache =
>       CacheBuilder.newBuilder()
>           .maximumSize(1000)
> ```

size can be increased to 10k

**Comment by raghuAtWings** [comment]
*[`420-delegate-service/src/main/java/io/harness/service/intfc/DelegateCache.java` line 28 at r2](https://reviewable.io/reviews/harness/harness-core/33951#-N3v_hOX1wau_2EU5aaA:-N3v_hOX1wau_2EU5aaB:b-vnwpgq) ([raw file](https://github.com/harness/harness-core/blob/f0ccd540460b694f166e417a51c7816d0436927e/420-delegate-service/src/main/java/io/harness/service/intfc/DelegateCache.java#L28)):*
> ```Java
>   void invalidateDelegateProfileCache(String accountId, String delegateProfileId);
> 
>   List<Delegate> getDelegates(String accountId, String delegateGroupId);
> ```

rename to getDelegatesForGroup

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33951)* status: :shipit: complete! ✓

**Review by raghuAtWings — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33951#-:-N3wCSkvCQL2L-um71Qs:bnfp4nl)* status: :shipit: complete! ✓

**Review by puneetsar — APPROVED** [review]
> Reviewed 2 of 7 files at r1, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/33951)* status: :shipit: complete! ✓

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java [hunk 1]
```diff
@@ -83,6 +84,21 @@ public DelegateProfile load(ImmutablePair<String, String> delegateProfileKey) {
             }
           });
 
+  private LoadingCache<ImmutablePair<String, String>, List<Delegate>> delegatesFromGroupCache =
+      CacheBuilder.newBuilder()
+          .maximumSize(10000)
```
Comment (abhinav-harness) [inline 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java:89]: cache size should be lesser 10k is too much for account id as we only have max 5k account per cluster and I don't cache holding data for all accounts too

Comment (jennyjames21) [reply 420-delegate-service/src/main/java/io/harness/service/impl/DelegateCacheImpl.java:89]: what's downside of having higher cache size?

**Pull request merged**